### PR TITLE
JS Ecosystem lecture notes

### PR DESCRIPTION
**Pull request by Hamled**
## Description
This is the lecture notes for a lecture on the modern JS build tools ecosystem.

## Feedback Requested
Is this too detailed?

I started explaining the various tools and quickly realized that the students would need some kind of understanding of a build system and the idea of building a software project in the first place, since we basically don't need to do it with Ruby or Rails projects.

But I may have ended up going into too much detail when explaining that background stuff, so let me know if it should be cut down (and what parts you think should be cut if you have any suggestions).

### CODE REVIEW

File: 10-JavaScript/js-ecosystem.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+- Understand the purpose of a build system for client-side JS projects
+- Understand the role of the following specific tools within a build system:
+    - [npm](https://npmjs.com/)
+    - [webpack](https://webpack.js.org/)
+    - [Babel](https://babeljs.io/)
+    - [webpack-dev-server](https://github.com/webpack/webpack-dev-server)
+- Explore a complex build system setup by "ejecting" the build configuration from a `create-react-app` project
+
+## What is a build system?
+A build system for a software project is set of one or more programs (generally referred to as tools) that are usually connected by "script" code (written in a general programming language or one specific to a particular build system) which directs those tools to transform the project's source code and data files into a complete, deployable "package" for end users.
```
Comment (kariabancroft) [inline 10-JavaScript/js-ecosystem.md]: "is a set of"

**Comment by Hamled** [comment]
@kariabancroft Yeah, I'm not sure how best to approach that.

If their goal is to simply include a third-party library for their project, that can generally be done without ejecting (e.g. I added the moment library to the react-timeline and react-chatlog apps that way).

On the other hand, if they need to customize their build process by adding some specific webpack loader or something else, that _would_ require ejecting. In that case the specific guidance gets a lot more complicated because the webpack config is already so complicated -- I'm not even sure how I would go about adding stuff into it.

**Comment by Hamled** [comment]
@tildeee I added a new definitions table and linked to it from the learning goals.

Let me know if you think I should re-arrange the content or otherwise do something to call out that the definitions and overview on build systems is the core of what they should be learning.

File: 10-JavaScript/js-ecosystem.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+  However, this approach has some downsides. Specifically, the browser is loading a _file_ URL which looks like `file:///Users/username/some/directory/to/a/file.html`, which is treated differently from an _http(s)_ URL. For large or potentially complex projects it's important to mimic the actual end-user experience as closely as possible so we really want the browser to access our site through HTTP just as it would any other website on the internet.
+
+  In order to accomplish this we need to run a web server which will actually provide our HTML, CSS, JavaScript, and other files from a given URL (generally `localhost:3000` or similar). That is the purpose of the webpack-dev-server tool, specifically built to integrate with a webpack-based build system.
+
+  In very complex applications the developers might modify this "dev server" to have custom behavior akin to an API, but for our purposes we just need to run it with the default setup. In this case the dev server can most closely be compared to running `rails server` for our Rails projects.
+
+## Create React App
```
Comment (tildeee) [inline 10-JavaScript/js-ecosystem.md]: Starting here, this section is for the following learning goal:
- Explore a complex build system setup by "ejecting" the build configuration from a `create-react-app` project

How would you feel if before this header, we had the following top-level header above it:
`# Explore A Complex Build System Setup through Create React App`

?

That might help structure a student's expectation for what to get out of this section

File: 10-JavaScript/js-ecosystem.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+Lastly, while the above webpack config from Create React App may have seemed really intense and unapproachable, rest assured that it is the result of many, many developers' efforts to build and maintain that config -- and to make it useful for as broad an audience as possible.
+
+In your personal projects, or your projects for work, it's very likely you can get away with a simpler configuration. And generally you will build that config piece-by-piece over time, adding in one new tool at a time when you end up needing it. Creating your build system in this incremental approach is vastly more approachable than starting with a huge one such as CRA's and trying to learn it all at once.
+
+## Definitions
+There's a lot of detail in the sections above, and while that can be helpful it's always good to have a quick reference. Consider the table below to be the core set of terms you should understand from this resource.
+
+| Term | Definition |
```
Comment (CheezItMan) [inline 10-JavaScript/js-ecosystem.md]: 👍

**Pull request merged**